pydash/pydash_app/impl/periodic_tasks.py

```diff
@@ -0,0 +1,395 @@
+        self.target(*args, **kwargs)
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __eq__(self, other):
+        return isinstance(other, _Task) and other.name == self.name
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__} name={self.name}, target={self.target}, next_run_dt={self.next_run_dt}>"
+
+
+class _TaskRemoval(_Task):
+    """
+    A placeholder '_Task' that will indicate to the scheduler
+    that the existing task that has `name` should be removed from the schedule.
+    """
+
+    def __init__(self, name):
+        super().__init__(name, None, None)
+
+
+class _BackgroundTask(_Task):
+    """
+    A task that is run only once, right after it is added to the scheduler.
+    """
+
+    def __init__(self, name, task):
+        """
+        :name: An identifier to find this task again later (and e.g. remove or alter it). Can be any hashable (using a string or a tuple of strings/integers is common.)
+        :target: A function (or other callable) that will perform this task's functionality.
+        """
+        super().__init__(name, task)
+
+
+class _PeriodicTask(_Task):
+    """
+    A task that is run many times periodically.
+    """
+
+    def __init__(self, name, task, interval, run_at_start=False):
+        """
+        :name: An identifier to find this task again later (and e.g. remove or alter it). Can be any hashable (using a string or a tuple of strings/integers is common.)
+        :target: A function (or other callable) that will perform this task's functionality.
+        :interval: A datetime.timedelta representing how frequently to run the given target.
+        :run_at_start: If true, runs task right after it was added to the scheduler, rather than only after the first interval has passed.
+        """
+        super().__init__(name, task)
+
+        if not isinstance(interval, datetime.timedelta):
+            raise ValueError(
+                f"`interval` is expected to be a `datetime.timedelta` instance, but `{interval}` is not."
+            )
+
+        self.interval = interval
+        if not run_at_start:
+            self.next_run_dt += interval
+
+    def update_for_next_run(self):
+        """
+        Makes the Task ready to be re-run once the next period has passed.
+        """
+        # A conscious choice has been made to always update next_run_dt w.r.t the current time.
+        # This means that even if a task's `next_run_dt` would lag behind,
+        # the task would not be run multiple times right after another.
+        self.next_run_dt = datetime.datetime.now() + self.interval
+        return self.next_run_dt
+
+    def __repr__(self):
+        """
+        Overrides superclass to show `interval` and `next_run_dt` in the string representation.
+        """
+        return f"<{self.__class__.__name__} name={self.name}, target={self.target}, interval={self.interval}, next_run_dt={self.next_run_dt}>"
+
+
+class TaskScheduler:
```

Comment:
I would argue for moving this into a new file

Reply:
Agreed.

Reply:
Yes, that is indeed nicer. :+1: